### PR TITLE
Support NotModified (304) for REST responses

### PR DESCRIPTION
**Pull request by cetanu**
There's currently no way to avoid sending a snapshot of configuration via a discovery response using REST, without Envoy considering it a failure.

For the situation where the configuration available on the control plane, and the configuration available on the envoy proxy are the same, a REST control plane must either:

1. Return a 200 OK with the full snapshot of configuration
2. Do something outside the norm, resulting in metrics being emitted that indicate an xDS update failure, in addition to warnings emitted to logs

I'd like to allow `HTTP 304 Not Modified` as a way of indicating that Envoy doesn't need to make any changes to configuration.

I don't really know C++ at all, so if this is completely wrong please let me know!
I hope that this change is okay.

### CODE REVIEW

**Comment by dnoe** [comment]
We're happy to help walk you through any C++ stuff. This change will want some change to a test, or a new test, to make sure the behavior works. Someone will get to reviewing this and will help you with that.

In the meantime, follow the instructions at https://github.com/envoyproxy/envoy/blob/master/CONTRIBUTING.md#fixing-dco to fix the DCO sign off.

**Comment by dnoe** [comment]
One more thing to get acquainted with, the instructions for auto fixing format: https://github.com/envoyproxy/envoy/blob/master/support/README.md#fixing-format-problems

**Comment by cetanu** [comment]
I fixed up the previous issues; in regards to writing a test for this, I could not find an existing test so I am quite lost on how to write one from scratch.

I can understand conceptually that I probably want to initialize the `RestApiFetcher` and then pass a `Http::MessagePtr&&` to the `onSuccess` method, which has a response status of 304, to see if it completes/doesn't raise an error.... but no idea how to do that

**Comment by dnoe** [comment]
I'll take this on - I'm happy to help you learn a bit about testing with gMock/google test.

It looks like those lines of code are tested by `//test/common/config:filesystem_subscription_impl_test`. You can run just this test with `bazel test //test/common/config:filesystem_subscription_impl_test`. You'll want to add to these tests some code that verifies when you send the response with 304 the expected things happen.

The part of this test code that actually sends the response with a particular status code looks to be in `test/common/config/http_subscription_test_harness.h`:

https://github.com/envoyproxy/envoy/blob/47dbc4f506274ad7dac9226fd947c01b9d243923/test/common/config/http_subscription_test_harness.h#L116

In the test harness, some objects are mocked out which means we can arbitrarily control what calls to their methods do, and also place expectations on certain methods being called with certain arguments.

This is a good primer on Googletest which will help you understand that test code: https://github.com/google/googletest/blob/master/googlemock/docs/Documentation.md

Hopefully this gives you some stuff to get started on and generate some questions for me and we will iterate from there.

File: source/common/http/rest_api_fetcher.cc [hunk 1]
```diff
@@ -30,7 +30,9 @@ void RestApiFetcher::initialize() { refresh(); }
 
 void RestApiFetcher::onSuccess(Http::MessagePtr&& response) {
   uint64_t response_code = Http::Utility::getResponseStatus(response->headers());
-  if (response_code != enumToInt(Http::Code::OK)) {
+  if (response_code == enumToInt(Http::Code::NotModified)) {
+    requestComplete();
```
Comment (lizan) [inline source/common/http/rest_api_fetcher.cc]: a `return;` here or it will parse the response?

Comment (cetanu) [reply source/common/http/rest_api_fetcher.cc]: Ah, yes it should return there.

**Comment by cetanu** [comment]
@dnoe I'm having some trouble getting this to build and test on OSX darwin:

```

rules_foreign_cc: Build failed!
rules_foreign_cc: Keeping temp build directory  and dependencies directory  for debug.
rules_foreign_cc: Please note that the directories inside a sandbox are still cleaned unless you specify '--sandbox_debug' Bazel command line flag.

rules_foreign_cc: Printing build logs:

_____ BEGIN BUILD LOGS _____
Bazel external C/C++ Rules #0.0.7. Building library 'zlib'
Environment:______________TMPDIR=/var/folders/m4/03w2g5951fzg8w7dsvn1qrv80000gn/T/
EXT_BUILD_ROOT=/private/var/tmp/_bazel_vsyrakis/c250969ecf5b606728407c2013db5bb1/sandbox/darwin-sandbox/2/execroot/envoy
INSTALLDIR=/private/var/tmp/_bazel_vsyrakis/c250969ecf5b606728407c2013db5bb1/sandbox/darwin-sandbox/2/execroot/envoy/bazel-out/darwin-fastbuild/genfiles/external/envoy/bazel/foreign_cc/zlib
__CF_USER_TEXT_ENCODING=0x1F5:0x0:0xF
PATH=/private/var/tmp/_bazel_vsyrakis/c250969ecf5b606728407c2013db5bb1/sandbox/darwin-sandbox/2/execroot/envoy:/usr/gnu/bin:/usr/local/bin:/bin:/usr/bin:.
BUILD_TMPDIR=/var/folders/m4/03w2g5951fzg8w7dsvn1qrv80000gn/T/tmp.F7uCP6D4
PWD=/private/var/tmp/_bazel_vsyrakis/c250969ecf5b606728407c2013db5bb1/sandbox/darwin-sandbox/2/execroot/envoy
EXT_BUILD_DEPS=/private/var/tmp/_bazel_vsyrakis/c250969ecf5b606728407c2013db5bb1/sandbox/darwin-sandbox/2/execroot/envoy/bazel_foreign_cc_deps_zlib
SHLVL=2
BUILD_LOG=bazel-out/darwin-fastbuild/genfiles/external/envoy/bazel/foreign_cc/zlib/logs/CMake.log
BUILD_SCRIPT=bazel-out/darwin-fastbuild/genfiles/external/envoy/bazel/foreign_cc/zlib/logs/CMake_script.sh
_=/usr/bin/env
__________________________-- The C compiler identification is AppleClang 10.0.0.10001145
-- Check for working C compiler: /var/tmp/_bazel_vsyrakis/c250969ecf5b606728407c2013db5bb1/sandbox/darwin-sandbox/2/execroot/envoy/external/local_config_cc/cc_wrapper.sh
-- Check for working C compiler: /var/tmp/_bazel_vsyrakis/c250969ecf5b606728407c2013db5bb1/sandbox/darwin-sandbox/2/execroot/envoy/external/local_config_cc/cc_wrapper.sh -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Looking for sys/types.h
-- Looking for sys/types.h - found
-- Looking for stdint.h
-- Looking for stdint.h - found
-- Looking for stddef.h
-- Looking for stddef.h - found
-- Check size of off64_t
-- Check size of off64_t - failed
-- Looking for fseeko
-- Looking for fseeko - found
-- Looking for unistd.h
-- Looking for unistd.h - found
-- Renaming
--     /var/tmp/_bazel_vsyrakis/c250969ecf5b606728407c2013db5bb1/sandbox/darwin-sandbox/2/execroot/envoy/external/com_github_madler_zlib/zconf.h
-- to 'zconf.h.included' because this file is included with zlib
-- but CMake generates it automatically in the build directory.
-- Configuring done
CMake Warning (dev):
  Policy CMP0042 is not set: MACOSX_RPATH is enabled by default.  Run "cmake
  --help-policy CMP0042" for policy details.  Use the cmake_policy command to
  set the policy and suppress this warning.

  MACOSX_RPATH is not specified for the following targets:

   zlib

This warning is for project developers.  Use -Wno-dev to suppress it.

-- Generating done
-- Build files have been written to: /var/folders/m4/03w2g5951fzg8w7dsvn1qrv80000gn/T/tmp.F7uCP6D4
[1/37] Building C object CMakeFiles/zlibstatic.dir/gzclose.o
[2/37] Building C object CMakeFiles/zlibstatic.dir/compress.o
[3/37] Building C object CMakeFiles/zlibstatic.dir/adler32.o
[4/37] Building C object CMakeFiles/zlibstatic.dir/gzlib.o
[5/37] Building C object CMakeFiles/zlibstatic.dir/crc32.o
[6/37] Building C object CMakeFiles/zlibstatic.dir/inftrees.o
[7/37] Building C object CMakeFiles/zlibstatic.dir/gzwrite.o
[8/37] Building C object CMakeFiles/zlibstatic.dir/gzread.o
[9/37] Building C object CMakeFiles/zlibstatic.dir/uncompr.o
[10/37] Building C object CMakeFiles/zlibstatic.dir/zutil.o
[11/37] Building C object CMakeFiles/zlibstatic.dir/inffast.o
[12/37] Building C object CMakeFiles/zlib.dir/compress.o
[13/37] Building C object CMakeFiles/zlib.dir/gzclose.o
[14/37] Building C object CMakeFiles/minigzip.dir/test/minigzip.o
[15/37] Building C object CMakeFiles/zlib.dir/adler32.o
[16/37] Building C object CMakeFiles/example.dir/test/example.o
[17/37] Building C object CMakeFiles/zlibstatic.dir/infback.o
[18/37] Building C object CMakeFiles/zlib.dir/crc32.o
[19/37] Building C object CMakeFiles/zlib.dir/gzlib.o
[20/37] Building C object CMakeFiles/zlib.dir/inftrees.o
[21/37] Building C object CMakeFiles/zlib.dir/gzwrite.o
[22/37] Building C object CMakeFiles/zlib.dir/gzread.o
[23/37] Building C object CMakeFiles/zlibstatic.dir/trees.o
[24/37] Building C object CMakeFiles/zlib.dir/uncompr.o
[25/37] Building C object CMakeFiles/zlib.dir/zutil.o
[26/37] Building C object CMakeFiles/zlib.dir/inffast.o
[27/37] Building C object CMakeFiles/zlib.dir/infback.o
[28/37] Building C object CMakeFiles/zlibstatic.dir/deflate.o
[29/37] Building C object CMakeFiles/zlibstatic.dir/inflate.o
[30/37] Linking C static library libz.a
FAILED: libz.a 
: && /usr/local/Cellar/cmake/3.13.4/bin/cmake -E remove libz.a && /usr/bin/libtool qc libz.a  CMakeFiles/zlibstatic.dir/adler32.o CMakeFiles/zlibstatic.dir/compress.o CMakeFiles/zlibstatic.dir/crc32.o CMakeFiles/zlibstatic.dir/deflate.o CMakeFiles/zlibstatic.dir/gzclose.o CMakeFiles/zlibstatic.dir/gzlib.o CMakeFiles/zlibstatic.dir/gzread.o CMakeFiles/zlibstatic.dir/gzwrite.o CMakeFiles/zlibstatic.dir/inflate.o CMakeFiles/zlibstatic.dir/infback.o CMakeFiles/zlibstatic.dir/inftrees.o CMakeFiles/zlibstatic.dir/inffast.o CMakeFiles/zlibstatic.dir/trees.o CMakeFiles/zlibstatic.dir/uncompr.o CMakeFiles/zlibstatic.dir/zutil.o && /usr/bin/ranlib libz.a && :
error: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/libtool: no output file specified (specify with -o output)
Usage: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/libtool -static [-] file [...] [-filelist listfile[,dirname]] [-arch_only arch] [-sacLT] [-no_warning_for_no_symbols]
Usage: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/libtool -dynamic [-] file [...] [-filelist listfile[,dirname]] [-arch_only arch] [-o output] [-install_name name] [-compatibility_version #] [-current_version #] [-seg1addr 0x#] [-segs_read_only_addr 0x#] [-segs_read_write_addr 0x#] [-seg_addr_table <filename>] [-seg_addr_table_filename <file_system_path>] [-all_load] [-noall_load]
[31/37] Building C object CMakeFiles/zlib.dir/trees.o
[32/37] Building C object CMakeFiles/zlib.dir/deflate.o
[33/37] Building C object CMakeFiles/zlib.dir/inflate.o
ninja: build stopped: subcommand failed.

_____ END BUILD LOGS _____
rules_foreign_cc: Build script location: bazel-out/darwin-fastbuild/genfiles/external/envoy/bazel/foreign_cc/zlib/logs/CMake_script.sh
rules_foreign_cc: Build log location: bazel-out/darwin-fastbuild/genfiles/external/envoy/bazel/foreign_cc/zlib/logs/CMake.log


Executed 0 out of 1 test: [0m[31m[1m1 fails to build[0m.

```

Tried to look up the solution to this and couldn't really find anything that would work :\

**Comment by dnoe** [comment]
Unfortunately I don't use OS X and nothing in that looks familiar. You should always be able to use the CI docker container to do a build, since it comes with it's own toolchain then your environment matters less. But, this is definitely slower performance. I would suggest asking in #envoy-dev on the Slack, which is quite active, and looks like there's also #envoy-osx-dev.

https://github.com/envoyproxy/envoy/blob/master/ci/README.md

**Comment by cetanu** [comment]
I can't seem to get this working on docker either, do you have an example of how you would run this test with docker?

**Comment by cetanu** [comment]
I'm in the process of running this inside a fresh ubuntu 18.04 container while following this guide:

https://github.com/envoyproxy/envoy/blob/master/bazel/README.md#quick-start-bazel-build-for-developers

**Comment by cetanu** [comment]
Okay I managed to run the tests with this dockerfile

```
FROM ubuntu:18.04

USER root

ENV BUILDIFIER_BIN=$GOPATH/bin/buildifier
ENV APT_KEY_DONT_WARN_ON_DANGEROUS_USAGE=1

RUN apt-get update && apt-get -y install software-properties-common curl

RUN add-apt-repository ppa:webupd8team/java
RUN echo "deb [arch=amd64] http://storage.googleapis.com/bazel-apt stable jdk1.8" | tee /etc/apt/sources.list.d/bazel.list
RUN echo debconf shared/accepted-oracle-license-v1-1 select true | debconf-set-selections
RUN echo debconf shared/accepted-oracle-license-v1-1 seen true | debconf-set-selections
RUN curl -s https://bazel.build/bazel-release.pub.gpg -o bazel.gpg && apt-key add bazel.gpg

RUN apt-get update && apt-get -y install \
   libtool \
   cmake \
   clang-format-7 \
   automake \
   autoconf \
   make \
   ninja-build \
   unzip \
   virtualenv \
   golang-go \
   git \
   openjdk-8-jdk \
   oracle-java8-installer \
   bazel

RUN go get -u github.com/bazelbuild/buildtools/buildifier
RUN echo 'build --action_env=PATH="/usr/local/bin:/opt/local/bin:/usr/bin:/bin"' >> $HOME/.bazelrc

ADD . /source
WORKDIR /source
```

However, if I was to add a test for this case, I would have no idea where to start. It doesn't seem like there's anything similar to what I need to do. The RestApiFetcher doesn't appear to be tested anywhere in that file you mentioned

**Comment by dnoe** [comment]
Circling back to this to hopefully get you some help with writing a test. I think what you want to do in the test is to send an xDS response with 304 Not Modified and verify that Envoy treats it as a success case (ie, doesn't increment the stats you discuss in your PR description) and deals correctly with the response body not including the update. Does that seem like a reasonable strategy to you? I'll take a look into the code in that test and come up with a write up that explains the general idea of how to accomplish that.

**Comment by TolikT** [comment]
Hello,
is it planned to be merged?

Best Regards,
Anatoli

**Comment by dnoe** [comment]
Someone definitely needs to add a test for this to be merged, if you @TolikT are also able to help @cetanu with writing a test help will be appreciated I am sure. Thanks.

**Comment by cetanu** [comment]
>  send an xDS response with 304 Not Modified and verify that Envoy treats it as a success case (ie, doesn't increment the stats you discuss in your PR description) and deals correctly with the response body not including the update

Are there any examples in the repo similar to this that I can look at?

I know this is what I need to do, I just have no idea how to do so. The file mentioned contains nothing that seems relevant for this case, and if it is, that just proves how incomprehensible it is for me.

**Comment by dnoe** [comment]
Sorry for being slow to respond on this, lots of other things on my plate. I'll try to dedicate a bit of time to this again. Ultimately what I think you'll want to do is to add functionality that allows arbitrary response code into `HttpSubscriptionTestHarness::DeliverConfigUpdate` where the 200 comes from. So add an argument with response code, and then send that response code where it's hardcoded to 200 today.

It might also be useful to chat on the Slack channel, although I think we're on opposite sides of the world so maybe asynchronous is better :)

@mergeconflict might also have more context on the xDS subscription test infrastructure and might be able to help too.

**Comment by mergeconflict** [comment]
> @mergeconflict might also have more context on the xDS subscription test infrastructure and might be able to help too.

I have a tiny bit of experience with this -- I wrote the `RdsAfterLdsInvalidated` test in `ads_integration_test.cc`, and I can chat on slack about how that works, if it helps...

**Comment by cetanu** [comment]
I have read these replies and I plan to action them as soon as I can. This week has been... challenging.

**Comment by cetanu** [comment]
about to push some commits to this branch as I'm not able to test locally

**Comment by cetanu** [comment]
Not sure how to reopen this... -_-

**Pull request closed without merging**